File: video-production/build_video.js
// Build the ZK-CID pitch video.
//
// Why this exists alongside build_video.js: the original pipeline burned
// subtitles converted from SRT, and FFmpeg's SRT path assumes a 384x288
// PlayRes. On a 1080p canvas that multiplied every FontSize by 3.75, so a
// "FontSize=30" caption rendered ~112px tall and covered the artwork. It also
// ran Ken Burns up to z=1.18, which crops 144px from each side and sliced the
// first letter off every card heading (their margins are only ~104px).
//
// This version renders captions from ASS with an explicit 1920x1080 PlayRes,
// caps the zoom at 1.04, and adds a fade-out per segment.
const fs = require("fs");
const path = require("path");
const { spawnSync } = require("child_process");

const BASE = __dirname;
const CARDS = path.join(BASE, "cards");
const SEGS = path.join(BASE, "segments");
const AUDIO = path.join(BASE, "audio");
const FFMPEG = process.env.FFMPEG_PATH || "ffmpeg";
const FPS = 30;
const W = 1920;
const H = 1080;
const MAX_ZOOM = 1.04;
const CRF_MASTER = "19";
const CRF_FINAL = "22";
const FADE_IN = 0.18;
const FADE_OUT = 0.3;

const SEGMENTS = [
  [path.join(CARDS, "card-01.png"), 15, "in"],
  [path.join(CARDS, "card-02.png"), 25, "out"],
  [path.join(CARDS, "card-03.png"), 30, "in"],
  [path.join(CARDS, "card-04.png"), 5, "out"],
  [path.join(BASE, "recordings", "combined-demo-60s.mp4"), 60, "video"],
  [path.join(CARDS, "card-05.png"), 25, "in"],
  [path.join(CARDS, "card-06.png"), 30, "out"],
  [path.join(CARDS, "card-07.png"), 20, "in"],
];

const VARIANTS = [
  { tag: "en-v2", srt: path.join(BASE, "subtitles_en.srt") },
  { tag: "zh-v2", srt: path.join(BASE, "subtitles_zh.srt") },
];

function run(args, cwd) {
  console.log(">>", args.slice(0, 6).join(" "), "...");
  const p = spawnSync(FFMPEG, args, { cwd, encoding: "utf8" });
  if (p.status !== 0) {
    console.error((p.stderr || "").slice(-4000));
    throw new Error("ffmpeg failed");
  }
}

function renderSegments() {
  fs.mkdirSync(SEGS, { recursive: true });
  const lines = [];

  SEGMENTS.forEach(([src, dur, mode], i) => {
    const out = path.join(SEGS, "v2-" + String(i + 1).padStart(2, "0") + ".mp4");
    const frames = Math.max(2, Math.round(dur * FPS));
    let vf;

    if (mode === "video") {
      vf = [
        "scale=1920:1080:force_original_aspect_ratio=decrease",
        "pad=1920:1080:(ow-iw)/2:(oh-ih)/2:color=0x050914",
        "fps=" + FPS,
        "fade=t=in:st=0:d=" + FADE_IN,
        "fade=t=out:st=" + (dur - FADE_OUT).toFixed(2) + ":d=" + FADE_OUT,
        "format=yuv420p",
      ].join(",");
      run([
        "-y", "-i", src, "-vf", vf,
        "-t", String(dur), "-r", String(FPS), "-an",
        "-c:v", "libx264", "-preset", "medium", "-crf", CRF_MASTER, out,
      ], BASE);
    } else {
      const step = (MAX_ZOOM - 1) / frames;
      const zexpr = mode === "in"
        ? "min(1.0+" + step.toFixed(8) + "*on," + MAX_ZOOM + ")"
        : "max(" + MAX_ZOOM + "-" + step.toFixed(8) + "*on,1.0)";
      const xexpr = "iw/2-(iw/zoom/2)";
      const yexpr = "ih/2-(ih/zoom/2)";
      vf = [
        "zoompan=z='" + zexpr + "':d=" + frames + ":x='" + xexpr + "':y='" + yexpr + "':s=" + W + "x" + H + ":fps=" + FPS,
        "eq=contrast=1.035:saturation=1.05",
        "unsharp=5:5:0.35:5:5:0.0",
        "fade=t=in:st=0:d=" + FADE_IN,
        "fade=t=out:st=" + (dur - FADE_OUT).toFixed(2) + ":d=" + FADE_OUT,
        "format=yuv420p",
      ].join(",");
      run([
        "-y", "-loop", "1", "-i", src, "-vf", vf,
        "-t", String(dur), "-r", String(FPS),
        "-c:v", "libx264", "-preset", "medium", "-crf", CRF_MASTER, out,
      ], BASE);
    }

    lines.push("file '" + path.basename(out) + "'");
  });

  const listFile = path.join(SEGS, "v2-list.txt");
  fs.writeFileSync(listFile, lines.join("\n"), "utf8");
  const concatOut = path.join(SEGS, "v2-concat.mp4");
  run(["-y", "-f", "concat", "-safe", "0", "-i", listFile, "-c", "copy", concatOut], SEGS);
  return concatOut;
}

function buildAss(srt, tag) {
  const ass = path.join(SEGS, "subs-" + tag + ".ass");
  const p = spawnSync(
    process.execPath,
    [path.join(BASE, "make_ass.js"), srt, ass],
    { encoding: "utf8" }
  );
  if (p.status !== 0) {
    console.error(p.stderr || "");
    throw new Error("subtitle conversion failed");
  }
  return ass;
}

function finalize(concatOut, srt, tag) {
  const ass = buildAss(srt, tag);
  const out = path.join(BASE, "zk-cid-pitch-video-" + tag + ".mp4");
  const wav = path.join(AUDIO, "ambient.wav");
  run([
    "-y", "-i", concatOut, "-i", wav,
    "-filter_complex",
    "[0:v]ass=" + path.basename(ass) + "[v];[1:a]volume=0.22,afade=t=in:st=0:d=2,afade=t=out:st=203:d=7[a]",
    "-map", "[v]", "-map", "[a]",
    "-c:v", "libx264", "-preset", "slow", "-crf", CRF_FINAL,
    "-pix_fmt", "yuv420p", "-movflags", "+faststart",
    "-c:a", "aac", "-b:a", "160k", "-shortest", out,
  ], SEGS);
  const mb = (fs.statSync(out).size / (1024 * 1024)).toFixed(1);
  console.log("FINAL:", out, mb, "MB");
}

const concatOut = renderSegments();
for (const variant of VARIANTS) {
  finalize(concatOut, variant.srt, variant.tag);
}

File: video-production/build_video.py
# -*- coding: utf-8 -*-
"""Build the ZK-CID hackathon pitch video:
1) generate ambient background music (numpy)
2) render each segment (card/screenshot) with Ken Burns + fades via ffmpeg
3) concat segments
4) burn subtitles + mix audio -> final MP4
"""
import os, subprocess, shutil
import numpy as np

BASE = r"E:\AI WORK\WEB 3.0 Decentralized\video-production"
CARDS = os.path.join(BASE, "cards")
SHOTS = r"E:\AI WORK\WEB 3.0 Decentralized\zk-cid\docs\assets"
SEGS = os.path.join(BASE, "segments")
AUDIO = os.path.join(BASE, "audio")

FFMPEG = r"C:\Users\Administrator\ffmpeg-extract\ffmpeg-9.0-essentials_build\bin\ffmpeg.exe"
FPS = 30
W, H = 1920, 1080

SEGMENTS = [
    (os.path.join(CARDS, "card-01.png"), 15, "in",  None),
    (os.path.join(CARDS, "card-02.png"), 25, "out", None),
    (os.path.join(CARDS, "card-03.png"), 30, "in",  None),
    (os.path.join(CARDS, "card-04.png"),  5, "out", None),
    (os.path.join(SHOTS, "demo-00-landing.png"),        12, "in",  "left"),
    (os.path.join(SHOTS, "demo-01-comparison.png"),     12, "out", "right"),
    (os.path.join(SHOTS, "demo-02-identity.png"),       12, "in",  None),
    (os.path.join(SHOTS, "demo-03-issued.png"),         12, "out", None),
    (os.path.join(SHOTS, "demo-03-issued-success.png"), 12, "in",  None),
    (os.path.join(CARDS, "card-05.png"), 25, "in",  None),
    (os.path.join(CARDS, "card-06.png"), 30, "out", None),
    (os.path.join(CARDS, "card-07.png"), 20, "in",  None),
]

def run(cmd, cwd=None):
    print(">>", " ".join(cmd[:7]), "...")
    p = subprocess.run(cmd, capture_output=True, text=True, cwd=cwd)
    if p.returncode != 0:
        print(p.stderr[-4000:])
        raise SystemExit("ffmpeg failed")
    return p

# ---------- 1. ambient music ----------
def make_music(path, total=216.0, sr=44100):
    rng = np.random.default_rng(7)
    n = int(total * sr)
    t = np.arange(n) / sr
    mix = np.zeros(n)
    chords = [
        [130.81, 164.81, 196.00, 246.94, 293.66],
        [110.00, 130.81, 164.81, 220.00, 246.94],
        [ 87.31, 130.81, 174.61, 220.00, 261.63],
        [ 98.00, 123.47, 146.83, 196.00, 246.94],
    ]
    seg = 8.0
    pos = 0.0
    while pos < total:
        chord = chords[int(pos // seg) % len(chords)]
        d = min(seg, total - pos)
        m = int(d * sr)
        tt = np.arange(m) / sr
        env = np.minimum(tt / 2.5, 1.0) * np.minimum((d - tt) / 1.5, 1.0)
        env = np.clip(env, 0, 1)
        acc = np.zeros(m)
        for f in chord:
            detune = 1 + rng.uniform(-0.0015, 0.0015)
            vib = 1 + 0.0015 * np.sin(2 * np.pi * 0.15 * tt)
            acc += np.sin(2 * np.pi * f * detune * vib * tt) * 0.5
            acc += 0.25 * np.sin(2 * np.pi * f * 2 * tt)
        acc += np.sin(2 * np.pi * (chord[0] / 2) * tt) * 0.9
        lfo = 0.5 + 0.5 * np.sin(2 * np.pi * 0.05 * tt)
        acc *= (0.25 + 0.75 * lfo)
        mix[int(pos * sr):int(pos * sr) + m] += acc * env
        pos += seg
    noise = rng.normal(0, 1, n)
    nf = np.convolve(noise, np.ones(512) / 512, mode="same")
    nenv = 0.5 + 0.5 * np.sin(2 * np.pi * 0.02 * t + 1.0)
    mix += nf * 0.05 * nenv
    mix = np.tanh(mix * 1.2)
    mix = mix / (np.max(np.abs(mix)) + 1e-9) * 0.35
    delay = int(0.012 * sr)
    right = np.roll(mix, delay)
    stereo = np.stack([mix, right], axis=1)
    pcm = (stereo * 32767).astype(np.int16)
    import wave
    with wave.open(path, "wb") as w:
        w.setnchannels(2); w.setsampwidth(2); w.setframerate(sr)
        w.writeframes(pcm.tobytes())
    print("music:", path, os.path.getsize(path) // 1024, "KB")

make_music(os.path.join(AUDIO, "ambient.wav"))

# ---------- 2. render segments ----------
concat_file = os.path.join(SEGS, "list.txt")
with open(concat_file, "w", encoding="utf-8") as f:
    for i, (img, dur, mode, pan) in enumerate(SEGMENTS, 1):
        out = os.path.join(SEGS, f"seg-{i:02d}.mp4")
        frames = int(dur * FPS)
        if mode == "in":
            zexpr = "min(zoom+0.0008,1.18)"
        else:
            zexpr = "max(1.15-0.0008*on,1.0)"
        if pan == "left":
            xexpr = "(iw-iw/zoom)/2-(iw/zoom)*0.04*(on/{})".format(frames)
            yexpr = "ih/2-(ih/zoom/2)"
        elif pan == "right":
            xexpr = "(iw-iw/zoom)/2+(iw/zoom)*0.04*(on/{})".format(frames)
            yexpr = "ih/2-(ih/zoom/2)"
        else:
            xexpr = "iw/2-(iw/zoom/2)"
            yexpr = "ih/2-(ih/zoom/2)"
        vf = (
            "scale=2560:1440:force_original_aspect_ratio=increase,"
            "crop=2560:1440,"
            "zoompan=z='{}':d={}:x={}:y={}:s={}x{}:fps={},"
            "format=yuv420p,"
            "fade=t=in:st=0:d=0.4,fade=t=out:st={:.1f}:d=0.5"
        ).format(zexpr, frames, xexpr, yexpr, W, H, FPS, dur - 0.5)
        cmd = [FFMPEG, "-y", "-loop", "1", "-i", img, "-vf", vf,
               "-t", str(dur), "-r", str(FPS),
               "-c:v", "libx264", "-preset", "medium", "-crf", "20", out]
        run(cmd)
        f.write("file '{}'\n".format(os.path.basename(out)))

concat_out = os.path.join(SEGS, "concat.mp4")
run([FFMPEG, "-y", "-f", "concat", "-safe", "0", "-i", concat_file,
     "-c", "copy", concat_out])

# ---------- 3. finalize variants ----------
def finalize(srt_abs, tag, font):
    if srt_abs is None:
        vf = None
    else:
        shutil.copyfile(srt_abs, os.path.join(SEGS, "subs.srt"))
        vf = ("subtitles=subs.srt:force_style="
              "'FontName={},FontSize=21,PrimaryColour=&H00FFFFFF,"
              "OutlineColour=&H80101420,BorderStyle=1,Outline=1.2,Shadow=0.6,MarginV=42'").format(font)
    out = os.path.join(BASE, "zk-cid-pitch-video-{}.mp4".format(tag))
    wav = os.path.join(AUDIO, "ambient.wav")
    if vf is None:
        run([FFMPEG, "-y", "-i", concat_out, "-i", wav,
             "-filter_complex",
             "[1:a]volume=0.30,afade=t=in:st=0:d=2,afade=t=out:st=208:d=6[a]",
             "-map", "0:v", "-map", "[a]",
             "-c:v", "copy", "-c:a", "aac", "-b:a", "192k", "-shortest", out],
            cwd=SEGS)
    else:
        run([FFMPEG, "-y", "-i", concat_out, "-i", wav,
             "-filter_complex",
             "[0:v]{}[v];[1:a]volume=0.30,afade=t=in:st=0:d=2,afade=t=out:st=208:d=6[a]".format(vf),
             "-map", "[v]", "-map", "[a]",
             "-c:v", "libx264", "-preset", "medium", "-crf", "19",
             "-c:a", "aac", "-b:a", "192k", "-shortest", out],
            cwd=SEGS)
    print("FINAL:", out, os.path.getsize(out) // (1024 * 1024), "MB")

finalize(os.path.join(BASE, "subtitles_en.srt"), "en", "Segoe UI")
finalize(os.path.join(BASE, "subtitles_zh.srt"), "zh", "Microsoft YaHei")
finalize(None, "nocap", "")
